File: src/game/projects/projects.ts
import { getSkillProgress } from '../activity/xp'
import type { EnchantmentRow, ProjectRow } from '../data/projectTypes'
import type { FacilityRow, GameDatabase } from '../data/types'
import { hasProjectKnowledge } from '../npcs/knowledge'
import { inventoryCount } from '../production/recipes'
import type { PlayerSave } from '../save/types'

export interface ProjectInput {
  itemId: string
  quantity: number
}

export interface SpecialProductionStation {
  facility: FacilityRow
  skillId: string
  skillName: string
  label: string
}

export function isCompleteProject(project: ProjectRow): boolean {
  if (project.Status === 'Needs Data') return false
  if (project.Instant !== 'Yes') return false
  if (typeof project['XP Reward'] !== 'number') return false
  if (typeof project['Output Quantity'] !== 'number') return false
  if (!project['Output Item / Target ID'] || !project['Facility ID'] || !project['Skill ID']) {
    return false
  }
  if (typeof project['Gold Cost'] !== 'number') return false
  return true
}

export function projectInputs(project: ProjectRow): ProjectInput[] {
  const out: ProjectInput[] = []
  const pairs: Array<[string | null, number | null]> = [
    [project['Input 1 Item ID'], project['Input 1 Quantity']],
    [project['Input 2 Item ID'], project['Input 2 Quantity']],
    [project['Input 3 Item ID'], project['Input 3 Quantity']],
    [project['Input 4 Item ID'], project['Input 4 Quantity']],
  ]
  for (const [itemId, quantity] of pairs) {
    if (itemId && typeof quantity === 'number' && quantity > 0) {
      out.push({ itemId, quantity })
    }
  }
  return out
}

export function projectSkillRequirements(
  project: ProjectRow,
): Array<{ skillId: string; level: number }> {
  const out: Array<{ skillId: string; level: number }> = []
  const pairs: Array<[string | null, number | null]> = [
    [project['Required Skill 1 ID'], project['Required Skill 1 Level']],
    [project['Required Skill 2 ID'], project['Required Skill 2 Level']],
    [project['Required Skill 3 ID'], project['Required Skill 3 Level']],
  ]
  for (const [skillId, level] of pairs) {
    if (skillId && typeof level === 'number') {
      out.push({ skillId, level })
    }
  }
  return out
}

export function getProject(db: GameDatabase, projectId: string): ProjectRow | undefined {
  return db.Projects.find((project) => project['Project ID'] === projectId)
}

export function getEnchantment(db: GameDatabase, enchantmentId: string): EnchantmentRow | undefined {
  return db.Enchantments.find((row) => row['Enchantment ID'] === enchantmentId)
}

export function isEnchantmentOutput(outputId: string): boolean {
  return outputId.startsWith('ENCH-')
}

export function meetsProjectSkills(save: PlayerSave, project: ProjectRow): boolean {
  return projectSkillRequirements(project).every(
    (requirement) => getSkillProgress(save, requirement.skillId).level >= requirement.level,
  )
}

export function meetsProjectKnowledge(
  db: GameDatabase,
  save: PlayerSave,
  project: ProjectRow,
): boolean {
  return hasProjectKnowledge(db, save, project['Skill ID']).ok
}

export function unmetProjectSkillRequirements(
  db: GameDatabase,
  save: PlayerSave,
  project: ProjectRow,
): Array<{ skillId: string; skillName: string; level: number; have: number }> {
  return projectSkillRequirements(project)
    .map((requirement) => {
      const have = getSkillProgress(save, requirement.skillId).level
      return {
        skillId: requirement.skillId,
        skillName:
          db.Skills.find((skill) => skill['Skill ID'] === requirement.skillId)?.['Display Name'] ??
          requirement.skillId,
        level: requirement.level,
        have,
      }
    })
    .filter((requirement) => requirement.have < requirement.level)
}

export function maxProjectsFromMaterials(save: PlayerSave, project: ProjectRow): number {
  const inputs = projectInputs(project)
  if (inputs.length === 0) return Number.POSITIVE_INFINITY
  let max = Number.POSITIVE_INFINITY
  for (const input of inputs) {
    max = Math.min(max, Math.floor(inventoryCount(save, input.itemId) / input.quantity))
  }
  return Number.isFinite(max) ? Math.max(0, max) : 0
}

export function maxProjectsFromGold(save: PlayerSave, project: ProjectRow): number {
  const cost = project['Gold Cost']
  if (cost <= 0) return Number.POSITIVE_INFINITY
  return Math.floor(save.gold / cost)
}

export function maxProjectQuantity(save: PlayerSave, project: ProjectRow): number {
  const materialMax = maxProjectsFromMaterials(save, project)
  const goldMax = maxProjectsFromGold(save, project)
  if (!Number.isFinite(materialMax) && !Number.isFinite(goldMax)) return 1
  return Math.max(0, Math.min(materialMax, goldMax))
}

/** Launch projects at a facility (skill gates affect completion, not listing). */
export function projectsForFacility(
  db: GameDatabase,
  facilityId: string,
  skillId?: string,
): ProjectRow[] {
  return db.Projects.filter(
    (project) =>
      isCompleteProject(project) &&
      project['Facility ID'] === facilityId &&
      (!skillId || project['Skill ID'] === skillId),
  ).sort((a, b) => {
    const aLevel = a['Required Skill 1 Level'] ?? 0
    const bLevel = b['Required Skill 1 Level'] ?? 0
    return aLevel - bLevel || a['Display Name'].localeCompare(b['Display Name'])
  })
}

/**
 * Special Production stations at a location, derived from Facilities + Projects.
 * Instant projects do not use the Primary Activity slot.
 */
export function specialProductionStationsAt(
  db: GameDatabase,
  locationId: string,
): SpecialProductionStation[] {
  const facilities = db.Facilities.filter(
    (facility) =>
      facility['Location ID'] === locationId &&
      facility.Status !== 'Needs Data' &&
      facility['Facility Type'] === 'Special Production Station',
  )

  const stations: SpecialProductionStation[] = []
  for (const facility of facilities) {
    const skillIds = new Set(
      db.Projects.filter(
        (project) =>
          isCompleteProject(project) && project['Facility ID'] === facility['Facility ID'],
      ).map((project) => project['Skill ID']),
    )
    for (const skillId of skillIds) {
      const skillName =
        db.Skills.find((skill) => skill['Skill ID'] === skillId)?.['Display Name'] ?? skillId
      stations.push({
        facility,
        skillId,
        skillName,
        label: skillName,
      })
    }
  }

  // Artisanry uses the Crafting Workshop facility (not typed as Special Production Station).
  const workshopProjects = db.Projects.filter(
    (project) =>
      isCompleteProject(project) &&
      project['Skill ID'] === 'SKL-0012' &&
      db.Facilities.some(
        (facility) =>
          facility['Facility ID'] === project['Facility ID'] &&
          facility['Location ID'] === locationId,
      ),
  )
  for (const project of workshopProjects) {
    const facility = db.Facilities.find((row) => row['Facility ID'] === project['Facility ID'])
    if (!facility) continue
    if (stations.some((station) => station.facility['Facility ID'] === facility['Facility ID'] && station.skillId === 'SKL-0012')) {
      continue
    }
    stations.push({
      facility,
      skillId: 'SKL-0012',
      skillName: 'Artisanry',
      label: 'Artisanry',
    })
  }

  return stations.sort((a, b) => a.label.localeCompare(b.label))
}

File: packages/ik_rules/lib/src/projects/projects.dart
import 'dart:math' as math;

import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../activity/xp.dart';
import '../activity/requirements.dart';
import '../equipment/specialist.dart';
import '../js_compat.dart';
import '../npcs/knowledge.dart';
import '../production/recipes.dart';
import '../save/generated/save_models.dart';

/// One material line of a project.
class ProjectInput {
  const ProjectInput({required this.itemId, required this.quantity});

  final String itemId;
  final num quantity;

  Map<String, Object?> toJson() => <String, Object?>{'itemId': itemId, 'quantity': quantity};
}

/// A skill gate on a project.
class ProjectSkillRequirement {
  const ProjectSkillRequirement({required this.skillId, required this.level});

  final String skillId;
  final num level;

  Map<String, Object?> toJson() => <String, Object?>{'skillId': skillId, 'level': level};
}

/// A Special Production station in the location menu.
class SpecialProductionStation {
  const SpecialProductionStation({
    required this.facility,
    required this.skillId,
    required this.skillName,
    required this.label,
  });

  final FacilityRow facility;
  final String skillId;
  final String skillName;
  final String label;

  Map<String, Object?> toJson() => <String, Object?>{
    'facilityId': facility.raw['Facility ID'],
    'skillId': skillId,
    'skillName': skillName,
    'label': label,
  };
}

/// Menu labels for Special Production stations (place names, not skill names).
String specialProductionStationLabel(String skillId, String skillName) {
  if (skillId == 'SKL-0011') return 'Smithing forge';
  if (skillId == 'SKL-0012') return 'Artisans workshop';
  if (skillId == 'SKL-0013') return 'Mages quarters';
  return skillName;
}

bool isCompleteProject(ProjectRow project) {
  if (project.raw['Status'] == 'Needs Data') return false;
  if (project.raw['Instant'] != 'Yes') return false;
  if (project.raw['XP Reward'] is! num) return false;
  if (project.raw['Output Quantity'] is! num) return false;
  if (isBlank(project.raw['Output Item / Target ID'] as String?) ||
      isBlank(project.raw['Facility ID'] as String?) ||
      isBlank(project.raw['Skill ID'] as String?)) {
    return false;
  }
  return project.raw['Gold Cost'] is num;
}

List<ProjectInput> projectInputs(ProjectRow project) {
  final out = <ProjectInput>[];
  for (var slot = 1; slot <= 4; slot += 1) {
    final itemId = project.raw['Input $slot Item ID'];
    final quantity = project.raw['Input $slot Quantity'];
    if (itemId is String && itemId.isNotEmpty && quantity is num && quantity > 0) {
      out.add(ProjectInput(itemId: itemId, quantity: quantity));
    }
  }
  return out;
}

/// Project inputs after equipped modifiers (Wizard's Hat essence discount).
List<ProjectInput> projectInputsForSave(PlayerSave save, ProjectRow project) {
  return projectInputs(project)
      .map(
        (input) => input.itemId == essenceItemId
            ? ProjectInput(itemId: input.itemId, quantity: wizardEssenceCost(input.quantity, save))
            : input,
      )
      .toList();
}

List<ProjectSkillRequirement> projectSkillRequirements(ProjectRow project) {
  final out = <ProjectSkillRequirement>[];
  for (var slot = 1; slot <= 3; slot += 1) {
    final skillId = project.raw['Required Skill $slot ID'];
    final level = project.raw['Required Skill $slot Level'];
    if (skillId is String && skillId.isNotEmpty && level is num) {
      out.add(ProjectSkillRequirement(skillId: skillId, level: level));
    }
  }
  return out;
}

ProjectRow? getProject(GameDatabase db, String projectId) {
  return db.projects.firstWhereOrNull((project) => project.raw['Project ID'] == projectId);
}

EnchantmentRow? getEnchantment(GameDatabase db, String enchantmentId) {
  return db.enchantments.firstWhereOrNull((row) => row.raw['Enchantment ID'] == enchantmentId);
}

bool isEnchantmentOutput(String outputId) => outputId.startsWith('ENCH-');

bool meetsProjectSkills(PlayerSave save, ProjectRow project) {
  return projectSkillRequirements(
    project,
  ).every((requirement) => getSkillProgress(save, requirement.skillId).level >= requirement.level);
}

bool meetsProjectKnowledge(GameDatabase db, PlayerSave save, ProjectRow project) {
  return hasProjectKnowledge(db, save, jsString(project.raw['Skill ID'])).ok;
}

/// A skill gate the player has not reached yet.
class UnmetProjectSkill {
  const UnmetProjectSkill({
    required this.skillId,
    required this.skillName,
    required this.level,
    required this.have,
  });

  final String skillId;
  final String skillName;
  final num level;
  final num have;

  Map<String, Object?> toJson() => <String, Object?>{
    'skillId': skillId,
    'skillName': skillName,
    'level': level,
    'have': have,
  };
}

List<UnmetProjectSkill> unmetProjectSkillRequirements(
  GameDatabase db,
  PlayerSave save,
  ProjectRow project,
) {
  return projectSkillRequirements(project)
      .map((requirement) {
        final displayName = db.skills
            .firstWhereOrNull((skill) => skill.raw['Skill ID'] == requirement.skillId)
            ?.raw['Display Name'];
        return UnmetProjectSkill(
          skillId: requirement.skillId,
          skillName: displayName is String ? displayName : requirement.skillId,
          level: requirement.level,
          have: getSkillProgress(save, requirement.skillId).level,
        );
      })
      .where((requirement) => requirement.have < requirement.level)
      .toList();
}

/// Crafts affordable from bag materials. Infinite when the project needs none.
num maxProjectsFromMaterials(PlayerSave save, ProjectRow project) {
  final inputs = projectInputsForSave(save, project);
  if (inputs.isEmpty) return double.infinity;
  num max = double.infinity;
  for (final input in inputs) {
    max = math.min(max, (inventoryCount(save, input.itemId) / input.quantity).floor());
  }
  return max.isFinite ? math.max(0, max) : 0;
}

/// Crafts affordable from gold. Infinite when the project is free.
num maxProjectsFromGold(PlayerSave save, ProjectRow project) {
  final cost = jsNumber(project.raw['Gold Cost'] ?? 0);
  if (cost <= 0) return double.infinity;
  return (save.gold / cost).floor();
}

num maxProjectQuantity(PlayerSave save, ProjectRow project) {
  final materialMax = maxProjectsFromMaterials(save, project);
  final goldMax = maxProjectsFromGold(save, project);
  if (!materialMax.isFinite && !goldMax.isFinite) return 1;
  return math.max(0, math.min(materialMax, goldMax));
}

/// Launch projects at a facility (skill gates affect completion, not listing).
List<ProjectRow> projectsForFacility(GameDatabase db, String facilityId, [String? skillId]) {
  final lookupId = projectFacilityIdForLookup(facilityId);
  final matching = db.projects
      .where(
        (project) =>
            isCompleteProject(project) &&
            project.raw['Facility ID'] == lookupId &&
            (isBlank(skillId) || project.raw['Skill ID'] == skillId),
      )
      .toList();
  mergeSort(
    matching,
    compare: (a, b) => jsCompareThen(
      jsNumberOrZero(a.raw['Required Skill 1 Level']) -
          jsNumberOrZero(b.raw['Required Skill 1 Level']),
      () => jsLocaleCompare(jsString(a.raw['Display Name']), jsString(b.raw['Display Name'])),
    ),
  );
  return matching;
}

/// Special Production stations at a location, derived from Facilities + Projects.
///
/// Instant projects do not use the Primary Activity slot.
List<SpecialProductionStation> specialProductionStationsAt(GameDatabase db, String locationId) {
  final facilities = db.facilities.where(
    (facility) =>
        facility.raw['Location ID'] == locationId &&
        facility.raw['Status'] != 'Needs Data' &&
        facility.raw['Facility Type'] == 'Special Production Station',
  );

  final stations = <SpecialProductionStation>[];
  for (final facility in facilities) {
    final lookupId = projectFacilityIdForLookup(jsString(facility.raw['Facility ID']));
    final skillIds = <String>{};
    for (final project in db.projects) {
      if (!isCompleteProject(project) || project.raw['Facility ID'] != lookupId) continue;
      skillIds.add(jsString(project.raw['Skill ID']));
    }
    for (final skillId in skillIds) {
      final displayName = db.skills
          .firstWhereOrNull((skill) => skill.raw['Skill ID'] == skillId)
          ?.raw['Display Name'];
      final skillName = displayName is String ? displayName : skillId;
      stations.add(
        SpecialProductionStation(
          facility: facility,
          skillId: skillId,
          skillName: skillName,
          label: specialProductionStationLabel(skillId, skillName),
        ),
      );
    }
  }

  // Artisanry uses the Crafting Workshop facility (not typed as Special Production Station).
  final workshops = db.facilities.where(
    (facility) =>
        facility.raw['Location ID'] == locationId &&
        projectFacilityIdForLookup(jsString(facility.raw['Facility ID'])) == 'FAC-0003',
  );
  for (final facility in workshops) {
    final lookupId = projectFacilityIdForLookup(jsString(facility.raw['Facility ID']));
    final hasArtisanry = db.projects.any(
      (project) =>
          isCompleteProject(project) &&
          project.raw['Skill ID'] == 'SKL-0012' &&
          project.raw['Facility ID'] == lookupId,
    );
    if (!hasArtisanry) continue;
    if (stations.any(
      (station) =>
          station.facility.raw['Facility ID'] == facility.raw['Facility ID'] &&
          station.skillId == 'SKL-0012',
    )) {
      continue;
    }
    stations.add(
      SpecialProductionStation(
        facility: facility,
        skillId: 'SKL-0012',
        skillName: 'Artisanry',
        label: specialProductionStationLabel('SKL-0012', 'Artisanry'),
      ),
    );
  }

  mergeSort(stations, compare: (a, b) => jsLocaleCompare(a.label, b.label));
  return stations;
}

/// Stations the player can see, hiding ones still gated by a quest.
List<SpecialProductionStation> specialProductionStationsVisibleAt(
  GameDatabase db,
  PlayerSave save,
  String locationId,
) {
  return specialProductionStationsAt(db, locationId)
      .where(
        (station) => entityVisibleForSave(
          db,
          save,
          'Facility',
          jsString(station.facility.raw['Facility ID']),
        ),
      )
      .toList();
}
